### PR TITLE
request parsing

### PR DESCRIPTION
**Pull request by mxsdev**
Support for request types, most of the logic is copied from `nextlove` w/ some tweaks

Because request has it overloaded, we will have to use `req.jsonBody` instead of `req.body`. Same goes for form data, which will go from `req.formData` to `req.formDataBody`. Happy to come up with better naming here.

### CODE REVIEW

**Review by codetheweb — APPROVED** [review]

File: src/types/route-spec.ts [hunk 1]
```diff
@@ -23,10 +23,10 @@ export type RouteSpec<AuthMiddlewares extends string> = {
   formData?: z.ZodObject<any>
   queryParams?: z.ZodObject<any>
   commonParams?: z.ZodObject<any>
+  urlEncodedFormData?: z.ZodObject<any>
```
Comment (codetheweb) [inline src/types/route-spec.ts:26]: is the other option `multipartFormData`? if so might want to rename `formData` to remove ambiguity

File: package.json [hunk 2]
```diff
@@ -71,6 +71,7 @@
     "clipanion": "^4.0.0-rc.3",
     "edge-runtime": "^2.5.7",
     "esbuild": "^0.19.11",
+    "form-urlencoded": "^6.1.4",
```
Comment (codetheweb) [inline package.json]: looks like this is still only used in tests

```suggestion
```

**Review by codetheweb — APPROVED** [review]

File: package.json (hunk 2 above)
Comment (mxsdev) [reply package.json]: yeah, hence why its a dev dependency....am i missing something?

Comment (codetheweb) [reply package.json]: this is in `dependencies` 😛 
Screenshot 2024-02-01 at 16 49 58

Comment (mxsdev) [reply package.json]: yeah, i fucked up the merge 🤦

Comment (mxsdev) [reply package.json]: i SWEAR it wasnt in there when i checked 

image

**Pull request merged**